README.md
# HaploPy – Haplotype estimation and phasing in Python

This package contains tools for estimating haplotype (or allele list) frequencies in a population using measurements of unphased genotype data, that is, phenotypes.

## Introduction

In layman terms, a phenotype is defined as an observation of two-allele sets
over multiple gene loci:

``` text
    Aa ––––––– Bb ––––––– Cc
    |          |          |
  locus 1    locus 2    locus 3
```

Note that the above datum doesn't reveal what are the exact haplotype (allele
sequence) pair behind the phenotype. Possible parent haplotype pairs that could
result into the above phenotype are given by

``` text
(ABC, abc), (aBC, Abc), (AbC, aBc), (abC, ABc), ...
```

In other words, the mapping that maps a haplotype pair to a phenotype is defined
by the example

``` text
(Abc, aBC) => (Aa, Bb, Cc)
```

and so on. Note that each item in the phenotype is a set of two alleles where the
order doesn't matter. 

**Problem:** Suppose that we have a set of phenotype observations from a large
population of N individuals. For each individual phenotype we would like to
estimate what are the most probable haplotype pair that resulted into the
phenotype. The main ingredient of solution is the estimation of individual
haplotype frequencies in the population 


## Installation

The package is found in PyPi.

``` shell
pip install haplopy
```

Alternatively, install development version manually using Pip:

``` bash
git clone https://github.com/malmgrek/haplopy.git
pip install -r requirements
pip install -e .
```

To check if the development version installed correctly, run tests with

``` shell
pytest -v 
```

## Examples

### Estimate haplotype frequencies

Simulate dataset using a prescribed haplotype probabilities and 
a multinomial distribution model.

``` python
import haplopy as hp


proba_haplotypes = {
    ("A", "B", "C"): 0.34,
    ("a", "B", "c"): 0.20,
    ("a", "B", "C"): 0.13,
    ("a", "b", "c"): 0.23,
    ("A", "b", "C"): 0.10
}

phenotypes = hp.multinomial.Model(proba_haplotypes).random(100)

fig = hp.plot.plot_haplotypes(proba_haplotypes)
fig = hp.plot.plot_phenotypes(phenotypes)
```

![Original relative haplotype frequencies](./doc/images/hinton-original.png "Original")

![Simulated phenotype observation set](./doc/images/bar.png "Phenotypes")

Pretend that we don't know the underlying haplotype distribution and let's try to estimate it.

``` python
model = hp.multinomial.Model().fit(phenotypes)
fig = hp.plot.plot_haplotypes(
    model.proba_haplotypes,
    thres=1.0e-6  # Hide probabilities smaller than this
)
```

![Estimated relative haplotype frequencies](./doc/images/hinton-estimated.png "Estimated")

### Phenotype phasing

Use an existing model to calculate the probabilities (conditional to given the
phenotype) of different diplotype representations of a given phenotype.

``` python
import haplopy as hp


model = hp.multinomial.Model({
    ("A", "B"): 0.4,
    ("A", "b"): 0.3,
    ("a", "B"): 0.2,
    ("a", "b"): 0.1
})

# A complete phenotype observation
model.calculate_proba_diplotypes(("Aa", "Bb"))
# {(('A', 'B'), ('a', 'b')): 0.4, (('A', 'b'), ('a', 'B')): 0.6}

# A phenotype with some missing SNPs
model.calculate_proba_diplotypes(("A.", ".."))
# {(('A', 'B'), ('A', 'B')): 0.17582417582417584,
#  (('A', 'B'), ('A', 'b')): 0.2637362637362637,
#  (('A', 'B'), ('a', 'B')): 0.17582417582417584,
#  (('A', 'B'), ('a', 'b')): 0.08791208791208792,
#  (('A', 'b'), ('A', 'b')): 0.09890109890109888,
#  (('A', 'b'), ('a', 'B')): 0.13186813186813184,
#  (('A', 'b'), ('a', 'b')): 0.06593406593406592}

```

In particular, phenotype phasing also enables computing the probabilities of 
different admissible phenotypes as well as imputation of missing data:

``` python
model.calculate_proba_phenotypes(("A.", ".."))
# {('AA', 'BB'): 0.17582417582417584,
#  ('AA', 'Bb'): 0.2637362637362637,
#  ('Aa', 'BB'): 0.17582417582417584,
#  ('Aa', 'Bb'): 0.21978021978021978,
#  ('AA', 'bb'): 0.09890109890109888,
#  ('Aa', 'bb'): 0.06593406593406592}

# Imputes with the most probable one
model.impute(("A.", ".."))
# ("AA", "Bb")
```



CHANGELOG.md
# Change Log


## 0.1.1

### Update
- Packaging to latest Setuptools style


## 0.1.0

### Add
- Multinomial model ML fitting using EM
- Multinomial model based phenotype phasing
- String matching for missing value expansions
- Imputing and phasing in presence of missing data


## X.Y.Z
### Add
### Change
### Fix
### Remove


haplopy/tests/test_main.py
from haplopy import reconstruct_parent_haplotyp

haplopy/tests/test_datautils.py
"""Unit tests for Datautils module

"""

from collections import Counter

import numpy as np
from numpy.testing import assert_equal
import pytest

from haplopy import datautils


@pytest.mark.parametrize("phenotypes,expected", [
    (
        [],
        []
    ),
    (
        [()],
        [()]
    ),
    (
        [("AA",)],
        [("A",)]
    ),
    (
        [("AA", "BB")],
        [("A", "B")]
    ),
    (
        [("Aa", "Bb")],
        [("a", "b"), ("a", "B"), ("A", "b"), ("A", "B")]
    ),
    (
        [("aA", "Bb", "Cc")],
        [
            ("a", "b", "C"), ("a", "b", "c"), ("a", "B", "C"), ("a", "B", "c"),
            ("A", "b", "C"), ("A", "b", "c"), ("A", "B", "c"), ("A", "B", "C")
        ]
    ),
    # Before parent haplotype ordering fixes (0.1.0), this example reproduces
    # an case with varying output ordering.
    (
        [
            ('aa', 'bb', 'cc'),
            ('Aa', 'Bb', 'Cc'),
            ('aa', 'bb', 'cc'),
            ('aa', 'bb', 'cc'),
            ('Aa', 'Bb', 'Cc'),
            ('aa', 'bb', 'cc'),
            ('Aa', 'Bb', 'Cc'),
            ('Aa', 'Bb', 'Cc'),
            ('Aa', 'Bb', 'Cc'),
            ('aa', 'bb', 'cc')
        ],
        [
            ('A', 'B', 'C'),
            ('A', 'B', 'c'),
            ('A', 'b', 'C'),
            ('A', 'b', 'c'),
            ('a', 'B', 'C'),
            ('a', 'B', 'c'),
            ('a', 'b', 'C'),
            ('a', 'b', 'c')
        ]
    )
])
def test_find_admissible_haplotypes(phenotypes, expected):
    res = datautils.find_admissible_haplotypes(phenotypes)
    assert set(res) == set(expected)  # Ordering doesn't matter
    return


@pytest.mark.parametrize("phenotype,expected", [
    (
        (),
        [((), ())]
    ),
    # One homozygous haplotype
    (
        ("AA",),
        [(("A",), ("A",))]
    ),
    (
        ("AA", "BB"),
        [(("A", "B"), ("A", "B"))]
    ),
    (
        ("Aa", "BB", "Cc"),
        [
            (("A", "B", "C"), ("a", "B", "c")),
            (("A", "B", "c"), ("a", "B", "C"))
        ]
    )
])
def test_factorize(phenotype, expected):
    res = datautils.factorize(phenotype)
    print(res)
    assert res == expected
    return


@pytest.mark.parametrize("diplotype_representation,haplotypes,expected", [
    (
        [],
        ["foobar", 1, 2],
        np.ndarray(shape=(3, 0), dtype=int)
    ),
    (
        [[(0, 0)]],
        [("A", "B")],
        np.array([[2]])
    ),
    (
        [[(0, 1), (3, 2)], [(2, 2)]],
        [("A", "b"), ("a", "B"), ("A", "B"), ("a", "b")],
        np.array([
            [1, 0, 0],
            [1, 0, 0],
            [0, 1, 2],
            [0, 1, 0]
        ])
    )
])
def test_build_diplotype_matrix(diplotype_representation, haplotypes, expected):
    res = datautils.build_diplotype_matrix(diplotype_representation, haplotypes)
    assert_equal(res.toarray(), expected)
    return


@pytest.mark.parametrize("haplotype,haplotypes,expected", [
    # Nothing to replace, present in reference list
    (
        ("A", "B"),
        [("A", "B"), ("A", "b"), ("a", "B"), ("a", "b")],
        [("A", "B")]
    ),
    # Nothing to replace, not present in reference list
    (
        ("x", "y"),
        [("A", "B"), ("A", "b"), ("a", "B"), ("a", "b")],
        []
    ),
    # Nothing to replace, empty reference list
    (
        ("A", "B"),
        [],
        []
    ),
    (
        (".", "b"),
        [("A", "B"), ("A", "b"), ("a", "B"), ("a", "b")],
        [("A", "b"), ("a", "b")]
    ),
    # A complete wildcard finds all
    (

        (".", "."),
        [("A", "B"), ("A", "b"), ("a", "B"), ("a", "b")],
        [("A", "B"), ("A", "b"), ("a", "B"), ("a", "b")]
    )
])
def test_match(haplotype, haplotypes, expected):
    res = datautils.match(haplotype, haplotypes)
    assert res == expected
    return


@pytest.mark.parametrize("haplotype", [
    ("A", "*"),
    ("A", "ö"),
    ("9", ",")
])
def test_match_invalid(haplotype):
    with pytest.raises(ValueError):
        datautils.match(haplotype, [("a", "b")])
    return


@pytest.mark.parametrize("diplotype,expected", [
    ((("A",), ("B")), ("AB",)),
    ((("A", "B"), ("A", "B")), ("AA", "BB")),
    ((("A", "B", "C"), ("a", "b", "c")), ("Aa", "Bb", "Cc")),
    ((("a", "b", "c"), ("A", "B", "C")), ("Aa", "Bb", "Cc")),
])
def test_unphase(diplotype, expected):
    res = datautils.unphase(diplotype)
    assert res == expected
    return


@pytest.mark.parametrize("diplotype,haplotypes,expected", [
    (
        (("A", "."), ("a", "b")),
        [("A", "B"), ("A", "b"), ("a", "B"), ("a", "b")],
        [(("A", "B"), ("a", "b")), (("A", "b"), ("a", "b"))]
    ),
    # A complete wildcard finds all
    (
        ((".", "."), (".", ".")),
        [("A", "B"), ("A", "b"), ("a", "B"), ("a", "b")],
        [
            (('A', 'B'), ('A', 'B')),
            (('A', 'B'), ('A', 'b')),
            (('A', 'B'), ('a', 'B')),
            (('A', 'B'), ('a', 'b')),
            (('A', 'b'), ('A', 'b')),
            (('A', 'b'), ('a', 'B')),
            (('A', 'b'), ('a', 'b')),
            (('a', 'B'), ('a', 'B')),
            (('a', 'B'), ('a', 'b')),
            (('a', 'b'), ('a', 'b')),
        ]
    ),
    # Unmatched come as they are
    (
        (("x", "y"), (".", "z")),
        [("A", "B"), ("A", "b"), ("a", "B"), ("a", "b")],
        [(("x", "y"), (".", "z"))]
    ),
    (
        (("a", "b"), ("A", "B")),
        [("x", "y")],
        [(("a", "b"), ("A", "B"))]
    ),
    (
        (("a", "b"), ("A", "B")),
        [],
        [(("a", "b"), ("A", "B"))]
    )
])
def test_fill(diplotype, haplotypes, expected):
    res = datautils.fill(diplotype, haplotypes)
    assert res == expected
    return


@pytest.mark.parametrize("counter,haplotypes,expected", [
    (
        Counter({("Aa", "Bb"): 2, ("AA", "BB"): 1}),
        [("A", "B"), ("A", "b"), ("a", "B"), ("a", "b")],
        [[(0, 3), (1, 2)], [(0, 0)]]
    )
    # Empty haplotypes
])
def test_build_diplotype_representation(counter, haplotypes, expected):
    res = datautils.build_diplotype_representation(counter, haplotypes)
    assert res == expected
    return


@pytest.mark.parametrize("phenotypes,expected", [
    (
        [],
        Counter()
    ),
    # Genotype permutation invariance
    (
        [("Aa", "Bb"), ("aA", "bB")],
        Counter({("Aa", "Bb"): 2})
    ),
    (
        [("AA", "BB", "CC"), ("Aa", "Bb", "cC")],
        Counter({("AA", "BB", "CC"): 1, ("Aa", "Bb", "Cc"): 1})
    )
])
def test_count_distinct(phenotypes, expected):
    assert datautils.count_distinct(phenotypes) == expected
    return


@pytest.mark.parametrize("phenotype,haplotypes,expected", [
    (
        ("a.", ".b"),
        [("A", "B"), ("a", "B"), ("A", "b"), ("a", "b")],
        [
            (('A', 'B'), ('a', 'b')),
            (('A', 'b'), ('a', 'B')),
            (('A', 'b'), ('a', 'b')),
            (('a', 'B'), ('a', 'b')),
            (('a', 'b'), ('a', 'b')),
        ]
    ),
    (
        ("aA", "Bb"),
        [("A", "B"), ("a", "B"), ("A", "b"), ("a", "b")],
        [(("A", "B"), ("a", "b")), (("A", "b"), ("a", "B"))]
    )
])
def test_factorize_fill(phenotype, haplotypes, expected):
    res = datautils.factorize_fill(phenotype, haplotypes)
    assert res == expected
    return


@pytest.mark.parametrize("x,expected", [
    ({"a": 2, "b": 42, "c": np.NaN}, ("b", 42)),
    ({"a": np.NaN}, (None, None))
])
def test_dmax(x, expected):
    assert datautils.dmax(x) == expected
    return


haplopy/tests/test_multinomial.py
"""Tests for the Multinomial module"""

from functools import reduce

import numpy as np
from numpy.testing import assert_equal, assert_almost_equal, assert_raises
import pytest

from haplopy.multinomial import log_multinomial, expectation_maximization, Model


np.random.seed(666)


def assert_dicts_almost_equal(x, y, **kwargs):
    (x_keys, x_values) = zip(*x.items())
    (y_keys, y_values) = zip(*y.items())
    assert x_keys == y_keys
    assert_almost_equal(x_values, y_values)


@pytest.mark.parametrize("args,expected", [
    ((1,), 0),
    # Note that this is not binomial(n, n)
    ((13, 13), np.log(10400600)),
    ((1, 2, 3, 4), np.log(12600)),
    # Large numbers
    ((666, 42, 313, 13), 849.87355176),
    # Invariant of argument permutations
    ((1, 2, 3, 4), log_multinomial(4, 2, 1, 3)),
    ((1, 2, 3, 4), log_multinomial(2, 1, 3, 4)),
    ((1, 2, 3, 4), log_multinomial(3, 4, 1, 2)),
    # Boundary values
    ((5, 0), 0),
    ((1, 2, 0, 4), log_multinomial(1, 2, 4))
])
def test_log_multinomial(args, expected):
    res = log_multinomial(*args)
    assert_almost_equal(res, expected)
    return


@pytest.mark.parametrize("kwargs,expected", [
    (
        {
            "phenotypes": [
                ("aa", "bb")
            ]
        },
        {
            ("a", "b"): 1.0
        }
    ),
    (
        {
            "phenotypes": [
                ("Aa", "Bb")
            ]
        },
        {
            ("A", "B"): 0.25,
            ("A", "b"): 0.25,
            ("a", "B"): 0.25,
            ("a", "b"): 0.25
        }
    ),
    (
        {
            "phenotypes": [
                ("aa", "BB"),
                ("AA", "BB")
            ]
        },
        {
            ("a", "B"): 0.5,
            ("A", "B"): 0.5
        }
    ),
    (
        {
            "phenotypes": [
                ("Aa", "BB", "Cc"),
                ("AA", "BB", "cc"),
                ("AA", "BB", "cc"),
                ("Aa", "BB", "Cc")
            ]
        },
        {
            ("a", "B", "c"): 0.0,
            ("A", "B", "C"): 0.0,
            ("A", "B", "c"): 0.75,
            ("a", "B", "C"): 0.25
        }
    )
])
def test_expectation_maximization(kwargs, expected):
    (res, _) = expectation_maximization(**kwargs)
    assert res == expected
    return


def test_model_init():
    proba_haplotypes = {("a", "b"): 0.5, ("A", "B"): 0.6}
    assert_raises(AssertionError, Model, proba_haplotypes)


@pytest.mark.parametrize("proba_haplotypes,n_obs,expected", [
    (
        {
            ("A", "B", "C"): 0.24,
            ("a", "b", "c"): 0.66,
            ("A", "b", "C"): 0.1
        },
        10,
        {
            ("A", "B", "C"): 0.25,
            ("a", "b", "c"): 0.75
        }
    ),
    (
        {
            ("A", "B", "C"): 0.24,
            ("a", "b", "c"): 0.66,
            ("A", "b", "C"): 0.1
        },
        1000,
        {
            ("A", "B", "C"): 0.2370,
            ("A", "b", "C"): 0.1045,


   ("a", "b", "c"): 0.6585
        }
    )
])
def test_model_fit(proba_haplotypes, n_obs, expected):
    model = Model(proba_haplotypes)
    phenotypes = model.random(n_obs)
    model_fitted = model.fit(phenotypes)
    # Assert that essentially nonzero probabilities coincide with expected
    result = {
        k: v for (k, v) in model_fitted.proba_haplotypes.items() if v >= 1e-8
    }
    assert_dicts_almost_equal(result, expected)
    return


@pytest.mark.parametrize("phenotype,fill_proba,expected", [
    # Leave complete phenotype unchanged
    (
        ("Aa", "Bb"),
        np.NaN,
        ("Aa", "Bb")
    ),
    # Unseen phenotype, remains unchanged
    (
        ("Aa", "xy"),
        np.NaN,
        ("Aa", "xy")
    ),
    (
        ("A.", "bB"),
        np.NaN,
        ("Aa", "Bb")
    ),
    (
        ("..", ".."),
        np.NaN,
        ("Aa", "BB")
    ),
    # Ignorant of genotype permutations
    (
        ("aA", ".."),
        np.NaN,
        ("Aa", "BB")
    )
])
def test_model_impute(phenotype, fill_proba, expected):
    # TODO: All options equally probable
    proba_haplotypes = {
        ("A", "B"): 0.4,
        ("a", "B"): 0.3,
        ("A", "b"): 0.2,
        ("a", "b"): 0.1
    }
    model = Model(proba_haplotypes)
    res = model.impute(phenotype, fill_proba=fill_proba)
    assert res == expected
    return


@pytest.mark.parametrize("proba_haplotypes,phenotypes,fill_probas,expected", [
    # All haplotypes present
    (
        {
            ("A", "B"): 0.1,
            ("a", "B"): 0.2,
            ("A", "b"): 0.3,
            ("a", "b"): 0.4
        },
        [
            ('Aa', 'Bb'),
            ('aa', 'bb'),
            ('Aa', 'bb'),
            ('Aa', 'BB'),
            ('aa', 'Bb'),
        ],
        [
            np.NaN,
            np.NaN,
            np.NaN,
            np.NaN,
            np.NaN
        ],
        [
            {(('A', 'B'), ('a', 'b')): 0.4, (('A', 'b'), ('a', 'B')): 0.6},
            {(('a', 'b'), ('a', 'b')): 1.0},
            {(('A', 'b'), ('a', 'b')): 1.0},
            {(('A', 'B'), ('a', 'B')): 1.0},
            {(('a', 'B'), ('a', 'b')): 1.0},
        ]
    ),
    # Some haplotypes missing from model
    (
        {
            ("A", "B"): 0.5,
            ("a", "b"): 0.5
        },
        [
            ('aa', 'bb'),
            ('Aa', 'Bb'),
            ('Aa', 'Bb'),
            ('AA', 'BB'),
            ('aa', 'bb'),
            ('AA', 'BB'),
        ],
        [
            np.NaN,
            np.NaN,
            0,
            np.NaN,
            np.NaN,
            np.NaN
        ],
        [
            {(('a', 'b'), ('a', 'b')): 1.0},
            {(('A', 'B'), ('a', 'b')): np.NaN, (('A', 'b'), ('a', 'B')): np.NaN},
            {(('A', 'B'), ('a', 'b')): 1.0, (('A', 'b'), ('a', 'B')): 0.0},
            {(('A', 'B'), ('A', 'B')): 1.0},
            {(('a', 'b'), ('a', 'b')): 1.0},
            {(('A', 'B'), ('A', 'B')): 1.0},
        ]
    ),
    # Diplotypes zero probability
    (
        {
            ("A", "B"): 0,
            ("a", "B"): 0.5,
            ("A", "b"): 0.5,
            ("a", "b"): 0
        },
        [
            ('aa', 'bb'),
            ("AA", "BB")
        ],
        [
            np.NaN,
            np.NaN
        ],
        [
            {(('a', 'b'), ('a', 'b')): 0},
            {(('A', 'B'), ('A', 'B')): 0},
        ]
    ),
    # Missing observations
    (
        {
            ("A", "B"): 0.1,
            ("a", "B"): 0.2,
            ("A", "b"): 0.3,
            ("a", "b"): 0.4
        },
        [
            ('A.', 'BB'),
            # Same but different order
            (".A", "BB"),
            ("..", "bb"),
            ("a.", ".b")
        ],
        [
            np.NaN,
            np.NaN,
            np.NaN,
            np.NaN
        ],
        [
            {(('A', 'B'), ('A', 'B')): 0.2, (('A', 'B'), ('a', 'B')): 0.8},
            {(('A', 'B'), ('A', 'B')): 0.2, (('A', 'B'), ('a', 'B')): 0.8},
            {
                (('A', 'b'), ('A', 'b')): .3**2/(.3**2+2*.3*.4+.4**2),
                (('A', 'b'), ('a', 'b')): 2*.3*.4/(.3**2+2*.3*.4+.4**2),
                (('a', 'b'), ('a', 'b')): .4**2/(.3**2+2*.3*.4+.4**2)
            },
            {
                (('A', 'B'), ('a', 'b')): 0.10526315,
                (('A', 'b'), ('a', 'B')): 0.15789473,
                (('A', 'b'), ('a', 'b')): 0.31578947,
                (('a', 'B'), ('a', 'b')): 0.21052631,
                (('a', 'b'), ('a', 'b')): 0.21052631,
            }
        ]
    ),
])
def test_proba_diplotypes(proba_haplotypes, phenotypes, fill_probas, expected):
    model = Model(proba_haplotypes)
    for (phenotype, p, e) in zip(phenotypes, fill_probas, expected):
        proba_diplotypes = model.calculate_proba_diplotypes(
            phenotype, fill_proba=p
        )
        assert_dicts_almost_equal(proba_diplotypes, e)
    return


@pytest.mark.parametrize("proba_haplotypes,phenotypes,fill_probas,expected", [
    # Missing observations
    (
        {
            ("A", "B"): 0.1,
            ("a", "B"): 0.2,
            ("A", "b"): 0.3,
            ("a", "b"): 0.4
        },
        [
            ('A.', 'BB'),
            # Same but different order
            (".A", "BB"),
            ("..", "bb"),
            ("a.", ".b"),
            # Impossible to impute
            ("a.", "xy"),
            (".a", "xy")
        ],
        [
            np.NaN,
            np.NaN,
            np.NaN,
            np.NaN,
            np.NaN,
            0
        ],
        [
            {("AA", "BB"): 0.2, ("Aa", "BB"): 0.8},
            {("AA", "BB"): 0.2, ("Aa", "BB"): 0.8},
            {
                ("AA", "bb"): .3**2/(.3**2+2*.3*.4+.4**2),
                ("Aa", "bb"): 2*.3*.4/(.3**2+2*.3*.4+.4**2),
                ("aa", "bb"): .4**2/(.3**2+2*.3*.4+.4**2)
            },
            {
                ("Aa", "Bb"): 0.2631578,
                ("Aa", "bb"): 0.31578947,
                ("aa", "Bb"): 0.21052631,
                ("aa", "bb"): 0.21052631,
            },
            # Result is ordered alphabetically
            {(".a", "xy"): np.NaN},
            {(".a", "xy"): 0}
        ]
    ),
])
def test_proba_phenotypes(proba_haplotypes, phenotypes, fill_probas, expected):
    model = Model(proba_haplotypes)
    for (phenotype, p, e) in zip(phenotypes, fill_probas, expected):
        proba_phenotypes = model.calculate_proba_phenotypes(
            phenotype, fill_proba=p
        )
        assert_dicts_almost_equal(proba_phenotypes, e)
    return


def test_model_serialize(tmp_path):
    d = tmp_path / "tmp"
    d.mkdir()
    p = d / "foobar.json"
    fp = str(p)
    model = Model({
        ("A", "B"): 0.4,
        ("A", "b"): 0.3,
        ("a", "B"): 0.2,
        ("a", "b"): 0.1
    })
    model.to_json(fp)
    assert model.proba_haplotypes == Model.from_json(fp).proba_haplotypes
    return


haplopy/multinomial.py
"""Multinomial distribution haplotype frequency inference model

TODO: Random generate EM initial values

"""


from __future__ import annotations

from functools import reduce
import json
import logging
from typing import Dict, List, Tuple

import numpy as np

from haplopy import datautils


def log_multinomial(*ns: int):
    """Logarithm of the multinomial coefficient

        (n1 + n2 + ... + nk)!
        ---------------------
        n1!  n2!  ...   nk!

    log(n!) = log(n) + log(n-1) + log(n-2) + ... + log(1)

    """

    def log_factorial(n):
        return np.log(np.arange(1, n + 1)).sum()

    return log_factorial(sum(ns)) - sum(map(log_factorial, ns))


def expectation_maximization(
        phenotypes: List[Tuple[str]],
        proba_haplotypes: Dict[Tuple[str], float]=None,
        max_iter: int=100,
        tol: float=1.0e-12
):
    """Expectation Maximization algorithm for haplotype probabilities

    Tries to iteratively find the Maximum Likelihood estimate.

    Philosophically, builds upon the hypotheses of Hardy-Weinberg equilibrium
    and random mating.

    Parameters
    ----------

    phenotypes : List[Tuple[str]]
        Observation data; list of phenotypes
    proba_haplotypes : Dict[Tuple[str], float]
        Initial haplotype probabilities
    max_iter : int
        Maximum number of iterations
    tol : float
        Stopping criterion with respect to Euclidean norm of probabilities vector

    """

    N = len(phenotypes)

    # If haplotype probabilities are not given, a uniform initial probability
    # is assumed, and only admissible haplotypes are considered
    if proba_haplotypes is None:
        haplotypes = datautils.find_admissible_haplotypes(phenotypes)
        Nh = len(haplotypes)
        probas = np.ones(Nh) / Nh
    else:
        (haplotypes, probas) = zip(*proba_haplotypes.items())

    counter = datautils.count_distinct(phenotypes)
    diplotype_representation = datautils.build_diplotype_representation(
        counter, haplotypes
    )

    # Diplotype count matrix for faster multiplication
    diplotype_matrix = datautils.build_diplotype_matrix(
        diplotype_representation, haplotypes
    )

    #
    # Separation into expectation and maximization
    #
    # This is not a mathematically accurate separation
    # but loosely based on [1]. The update formulae are non-trivial to
    # derive. A clear mathematical proof is found, e.g., in [2].
    #
    # [1] J. Polanska, The EM algorithm and its implementation for the
    #     estimation of frequencies of SNP-haplotypes, Int. J. Appl. Math.
    #     Comput. Sci., 2003, Vol. 13, No. 3, 419-429.
    #
    # [2] Mikko Koivisto, 582673 Computational Genotype Analysis (lecture 2
    #     notes), Uni. Helsinki
    #

    def expectation(probas: np.ndarray):
        """Calculate probability for each pair and evaluate log-likelihood

        """

        def calculate(inds):
            return np.array([
                2 ** (i != j) * probas[i] * probas[j] for (i, j) in inds
            ])

        Ps_raws = [calculate(inds) for inds in diplotype_representation]
        Ps_units = np.hstack([
            Ps * n / Ps.sum() / N for (Ps, n) in zip(Ps_raws, counter.values())
        ])
        log_likelihood = sum([
            n * np.log(Ps.sum()) for (Ps, n) in zip(Ps_raws, counter.values())
        ]) + log_multinomial(*counter.values())

        return (Ps_units, log_likelihood)

    def maximization(Ps):
        """Calculate next estimates of haplotype probabilities

        """
        return 0.5 * diplotype_matrix.dot(Ps)

    n_iter = 0
    step = np.inf
    logging.info("Start EM algorithm")
    while (n_iter <= max_iter) and (step > tol):
        (Ps, log_likelihood) = expectation(probas)
        delta = maximization(Ps) - probas
        step = np.linalg.norm(delta)
        probas = probas + delta
        logging.info(
            "Iteration {0} | log(L) = {1:.4e} | step = {2: .4e}".format(
                n_iter, log_likelihood, step
            )
        )
        n_iter += 1

    proba_haplotypes = dict(zip(haplotypes, probas))

    return (proba_haplotypes, log_likelihood)


class Model():
    """Haplotype multinomial distribution estimator

    Examples
    --------
    >>> model = Model({("a", "b"): 0.5, ("A", "B"): 0.5})
    >>> phenotypes = model.random(10)
    >>> model_fitted = Model.fit(phenotypes)
    >>> model_fitted.proba_haplotypes
    {('A', 'B'): 0.6,
     ('A', 'b'): 0.0,
     ('a', 'B'): 0.0,
     ('a', 'b'): 0.4}


    """

    def __init__(self, proba_haplotypes: Dict[Tuple[str], float]):
        (haplotypes, probas) = zip(*proba_haplotypes.items())
        assert abs(sum(probas) - 1) < 1e-8, "Probabilities must sum to one"
        self.proba_haplotypes = proba_haplotypes
        return

    def random(self, n_obs: int) -> List[Tuple[str]]:
        """Random generate phenotypes

        """
        (haplotypes, probas) = zip(*self.proba_haplotypes.items())
        parent_inds = np.dot(
            np.random.multinomial(1, probas, 2 * n_obs),
            np.arange(len(probas))
        ).reshape(n_obs, 2)
        return [
            tuple(
                "".join(sorted(snp)) for snp in zip(haplotypes[i], haplotypes[j])
            )
            for (i, j) in parent_inds
        ]

    @classmethod
    def fit(cls, phenotypes: Tuple[str], **kwargs) -> Model:
        """Fit maximum likelihood haplotype probabilities using EM algorithm

        Implemented as a classmethod so that we can use given phenotypes
        to initialize the model conveniently.

        """
        (proba_haplotypes, _) = expectation_maximization(phenotypes, **kwargs)
        return cls(proba_haplotypes=proba_haplotypes)

    def calculate_proba_diplotypes(
            self,
            phenotype: Tuple[str],
            fill_proba: float=np.NaN,
    ) -> Dict[Tuple[int], float]:
        """Calculate admissible diplotypes' conditional probabilities

        Examples
        --------
        >>> model = Model({("a", "b"): 0.2, ("A", "B"): 0.5, ("a", "B"): 0.3})
        >>> model.calculate_proba_diplotypes(("Aa", "B."))
        {(('A', 'B'), ('a', 'B')): 0.6, (('A', 'B'), ('a', 'b')): 0.4}

        """

        # Extended diplotypes where missing data is filled if possible
        #
        # At the moment we have decided not to implement a tailored data type
        # for diplotypes, we need to use this complicated and messy composite
        # function to try to ensure there are no duplicate permutations in the
        # filled list. If we had a hashable and symmetric data type for
        # diplotypes, we could just use set (or frozenset) to trivially reduce
        # to unique diplotypes.
        diplotypes = datautils.factorize_fill(phenotype, self.proba_haplotypes)

        # Model's haplotypes might not contain all of the constituent haplotypes
        # of the given set of phenotypes. The probability of such haplotypes
        # will be considered `fill_proba`. Note that the implied ambiguity in
        # unit summability of the NaN-probability distribution doesn't ruin the
        # calculation of diplotype probabilities (with the existing haplotypes)
        # because each conditional probability is normalized.
        proba_haplotypes = {
            h: self.proba_haplotypes.get(h, fill_proba)
            for diplotype in diplotypes for h in diplotype
        }

        def calculate(ds):
            return np.array([
                2 ** (h1 != h2) * proba_haplotypes[h1] * proba_haplotypes[h2]
                for (h1, h2) in ds
            ])

        def normalize(x):
            return x / x.sum() if any(x) else x

        def to_dict(ds):
            return dict(zip(ds, normalize(calculate(ds))))

        return to_dict(diplotypes)

    def calculate_proba_phenotypes(self, phenotype: Tuple[str], **kwargs):
        """Compute probabilities of different options that fill a given phenotype

        Examples
        --------
        >>> model = Model({("a", "b"): 0.2, ("A", "B"): 0.5, ("a", "B"): 0.3})
        >>> model.calculate_proba_phenotypes(("Aa", ".."))
        {("Aa", "BB"): 0.6, ("Aa", "Bb": 0.4)
        """
        proba_diplotypes = self.calculate_proba_diplotypes(phenotype, **kwargs)

        def agg(cum, xs):
            (diplotype, proba) = xs
            phenotype = datautils.unphase(diplotype)
            # There can be multiple different diplotypes that result into the
            # same phenotype
            return (
                {**cum, **{phenotype: proba}} if phenotype not in cum
                else {**cum, **{phenotype: proba + cum[phenotype]}}
            )

        return reduce(agg, proba_diplotypes.items(), dict())

    def impute(self, phenotype: Tuple[str], **kwargs):
        """Impute by selecting the most probable diplotype

        Examples
        --------
        >>> model = Model({("a", "b"): 0.2, ("A", "B"): 0.5, ("a", "B"): 0.3})
        >>> model.impute(("Aa", "BB"))

        """
        proba_phenotypes = self.calculate_proba_phenotypes(phenotype)
        (argmax, m) = datautils.dmax(proba_phenotypes)
        return argmax if argmax else phenotype

    def to_json(self, fp):
        """Save existing model to hard disk

        """

        def jsonify(x):
            return {"".join(k): v for (k, v) in x.items()}

        with open(fp, "w+") as f:
            json.dump(jsonify(self.proba_haplotypes), f)

        return

    @classmethod
    def from_json(cls, fp):
        """Instantiate a new model from a JSON on disk

        """

        def unjsonify(x):
            return {tuple(k): v for (k, v) in x.items()}

        with open(fp, "r") as f:
            raw = json.load(f)

        return cls(proba_haplotypes=unjsonify(raw))
